### ISSUE TITLE
JavaScript error in Normalization

### ISSUE THREAD

**Issue by ibennetch** [phpmyadmin/phpmyadmin#12927]
See also http://stackoverflow.com/questions/41808961/a-fatal-javascript-error

I get a JavaScript error when attempting 3NF normalization of a table.

It doesn't matter what I pick for the first two steps, I can skip them/tell phpMyAdmin that everything is already optimized. For the third step, "Step 3.1 Find transitive dependencies", a checkbox must be selected, then a JavaScript error is reported. This was mine, tested with QA_4_6 (so basically 4.6.6) and master on the demo server:

```
{
    "pma_version": "4.6.6",
    "browser_name": "FIREFOX",
    "browser_version": "50.0",
    "user_os": "Mac",
    "server_software": "nginx/1.6.2",
    "user_agent_string": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:50.0) Gecko/20100101 Firefox/50.0",
    "locale": "en",
    "configuration_storage": "enabled",
    "php_version": "7.0.15-1+0~20170120094752.20+jessie~1.gbpe03972",
    "exception_type": "js",
    "exception": {
        "mode": "stack",
        "name": "SyntaxError",
        "message": "JSON.parse: unexpected character at line 1 column 1 of the JSON data",
        "stack": [
            {
                "func": "n.parseJSON",
                "line": 4,
                "column": "5492",
                "context": [
                    "/*! jQuery v2.1.4 | (c) 2005, 2015 jQuery Foundation, Inc. | jquery.org/lic//...",
                    "!function(a,b){\"object\"==typeof module&&\"object\"==typeof module.exports?mod//...",
                    "return M.access(a,b,c)},removeData:function(a,b){M.remove(a,b)},_data:funct//...",
                    "void 0===c?d&&\"get\"in d&&null!==(e=d.get(a,b))?e:(e=n.find.attr(a,b),null==//...",
                    ";",
                    "",
                    "function sprintf() {",
                    "/*",
                    " * Copyright (c) 2013 Kevin van Zonneveld (http://kvz.io)"
                ],
                "filename": "jquery/jquery-2.1.4.min.js"
            },
            {
                "func": "goTo3NFStep2/<.success",
                "line": 360,
                "column": "31",
                "context": [
                    "            type: \"GET\",",
                    "            url: \"normalization.php\",",
                    "            data: datastring,",
                    "            async:false,",
                    "            success: function(data) {",
                    "                data_parsed = $.parseJSON(data.message);",
                    "                if (data.success === true) {",
                    "                    extra += data_parsed.html;",
                    "                } else {",
                    "                    PMA_ajaxShowMessage(data.error, false);",
                    "                }"
                ],
                "filename": "normalization.js"
            },
            {
                "func": "n.Callbacks/j",
                "line": 2,
                "column": "26920",
                "context": [
                    "/*! jQuery v2.1.4 | (c) 2005, 2015 jQuery Foundation, Inc. | jquery.org/lic//...",
                    "!function(a,b){\"object\"==typeof module&&\"object\"==typeof module.exports?mod//...",
                    "return M.access(a,b,c)},removeData:function(a,b){M.remove(a,b)},_data:funct//...",
                    "void 0===c?d&&\"get\"in d&&null!==(e=d.get(a,b))?e:(e=n.find.attr(a,b),null==//...",
                    ";",
                    "",
                    "function sprintf() {"
                ],
                "filename": "jquery/jquery-2.1.4.min.js"
            },
            {
                "func": "n.Callbacks/k.fireWith",
                "line": 2,
                "column": "27738",
                "context": [
                    "/*! jQuery v2.1.4 | (c) 2005, 2015 jQuery Foundation, Inc. | jquery.org/lic//...",
                    "!function(a,b){\"object\"==typeof module&&\"object\"==typeof module.exports?mod//...",
                    "return M.access(a,b,c)},removeData:function(a,b){M.remove(a,b)},_data:funct//...",
                    "void 0===c?d&&\"get\"in d&&null!==(e=d.get(a,b))?e:(e=n.find.attr(a,b),null==//...",
                    ";",
                    "",
                    "function sprintf() {"
                ],
                "filename": "jquery/jquery-2.1.4.min.js"
            },
            {
                "func": "x",
                "line": 4,
                "column": "11251",
                "context": [
                    "/*! jQuery v2.1.4 | (c) 2005, 2015 jQuery Foundation, Inc. | jquery.org/lic//...",
                    "!function(a,b){\"object\"==typeof module&&\"object\"==typeof module.exports?mod//...",
                    "return M.access(a,b,c)},removeData:function(a,b){M.remove(a,b)},_data:funct//...",
                    "void 0===c?d&&\"get\"in d&&null!==(e=d.get(a,b))?e:(e=n.find.attr(a,b),null==//...",
                    ";",
                    "",
                    "function sprintf() {",
                    "/*",
                    " * Copyright (c) 2013 Kevin van Zonneveld (http://kvz.io)"
                ],
                "filename": "jquery/jquery-2.1.4.min.js"
            },
            {
                "func": ".send/b/<",
                "line": 4,
                "column": "14765",
                "context": [
                    "/*! jQuery v2.1.4 | (c) 2005, 2015 jQuery Foundation, Inc. | jquery.org/lic//...",
                    "!function(a,b){\"object\"==typeof module&&\"object\"==typeof module.exports?mod//...",
                    "return M.access(a,b,c)},removeData:function(a,b){M.remove(a,b)},_data:funct//...",
                    "void 0===c?d&&\"get\"in d&&null!==(e=d.get(a,b))?e:(e=n.find.attr(a,b),null==//...",
                    ";",
                    "",
                    "function sprintf() {",
                    "/*",
                    " * Copyright (c) 2013 Kevin van Zonneveld (http://kvz.io)"
                ],
                "filename": "jquery/jquery-2.1.4.min.js"
            },
            {
                "func": ".send",
                "line": 4,
                "column": "14951",
                "context": [
                    "/*! jQuery v2.1.4 | (c) 2005, 2015 jQuery Foundation, Inc. | jquery.org/lic//...",
                    "!function(a,b){\"object\"==typeof module&&\"object\"==typeof module.exports?mod//...",
                    "return M.access(a,b,c)},removeData:function(a,b){M.remove(a,b)},_data:funct//...",
                    "void 0===c?d&&\"get\"in d&&null!==(e=d.get(a,b))?e:(e=n.find.attr(a,b),null==//...",
                    ";",
                    "",
                    "function sprintf() {",
                    "/*",
                    " * Copyright (c) 2013 Kevin van Zonneveld (http://kvz.io)"
                ],
                "filename": "jquery/jquery-2.1.4.min.js"
            },
            {
                "func": ".ajax",
                "line": 4,
                "column": "10710",
                "context": [
                    "/*! jQuery v2.1.4 | (c) 2005, 2015 jQuery Foundation, Inc. | jquery.org/lic//...",
                    "!function(a,b){\"object\"==typeof module&&\"object\"==typeof module.exports?mod//...",
                    "return M.access(a,b,c)},removeData:function(a,b){M.remove(a,b)},_data:funct//...",
                    "void 0===c?d&&\"get\"in d&&null!==(e=d.get(a,b))?e:(e=n.find.attr(a,b),null==//...",
                    ";",
                    "",
                    "function sprintf() {",
                    "/*",
                    " * Copyright (c) 2013 Kevin van Zonneveld (http://kvz.io)"
                ],
                "filename": "jquery/jquery-2.1.4.min.js"
            },
            {
                "func": "goTo3NFStep2",
                "line": 354,
                "column": "9",
                "context": [
                    "            \"ajax_request\": true,",
                    "            \"db\": PMA_commonParams.get('db'),",
                    "            \"tables\": JSON.stringify(tablesTds),",
                    "            \"pd\": JSON.stringify(pd),",
                    "            \"getNewTables3NF\":1};",
                    "        $.ajax({",
                    "            type: \"GET\",",
                    "            url: \"normalization.php\",",
                    "            data: datastring,",
                    "            async:false,",
                    "            success: function(data) {"
                ],
                "filename": "normalization.js"
            },
            {
                "func": "processDependencies",
                "line": 422,
                "column": "9",
                "context": [
                    "            }",
                    "        }",
                    "    });",
                    "    backup = $(\"#mainContent\").html();",
                    "    if (isTransitive === true) {",
                    "        goTo3NFStep2(pd, tablesTds);",
                    "    } else {",
                    "        goTo2NFStep2(pd, primary_key);",
                    "    }",
                    "    return false;",
                    "}"
                ],
                "filename": "normalization.js"
            },
            {
                "func": "onclick",
                "line": "1",
                "column": "1",
                "uri": "normalization.php?goto=tbl_structure.php&back=tbl_structure.php",
                "scriptname": "normalization.php"
            }
        ],
        "uri": "index.php?"
    },
    "script_name": "index.php",
    "microhistory": null
```

**Comment by ibennetch** [comment]
The Stack Overflow reporter's error message is:
```
{
"pma_version": "4.6.5.2",
"browser_name": "CHROME",
"browser_version": "51.0.2704.79",
"user_os": "Win",
"server_software": "Apache/2.4.23 (Win64) PHP/7.0.10",
"user_agent_string": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.79 Safari/537.36 Edge/14.14393",
"locale": "en",
"configuration_storage": "disabled",
"php_version": "7.0.10",
"exception_type": "js",
"exception": {
    "mode": "stack",
    "name": "TypeError",
    "message": "Unable to get property 'newTables' of undefined or null reference",
    "stack": [
        {
            "line": 235,
            "func": "",
            "context": [
                "                lastException = null;",
                "                notifyHandlers.apply(null, [stack, null].concat(args));",
                "            }",
                "        }, (stack.incomplete ? 2000 : 0));",
                "",
                "        throw ex; // re-throw to propagate to the top level (and cause wind//...",
                "    }",
                "",
                "    report.subscribe = subscribe;",
                "    report.unsubscribe = unsubscribe;",
                "    return report;"
            ],
            "column": "",
            "filename": "tracekit/tracekit.js"
        },
        {
            "func": "new_func",
            "line": 249,
            "column": "21",
            "context": [
                "     */",
                "    wrap_function: function (func) {",
                "        if (!func.wrapped) {",
                "            var new_func = function () {",
                "                try {",
                "                    return func.apply(this, arguments);",
                "                } catch (x) {",
                "                    TraceKit.report(x);",
                "                }",
                "            };",
                "            new_func.wrapped = true;"
            ],
            "filename": "error_report.js"
        },
        {
            "func": "n.event.dispatch",
            "line": 3,
            "column": "6410",
            "context": [
                "/*! jQuery v2.1.4 | (c) 2005, 2015 jQuery Foundation, Inc. | jquery.org/lic//...",
                "!function(a,b){\"object\"==typeof module&&\"object\"==typeof module.exports?mod//...",
                "return M.access(a,b,c)},removeData:function(a,b){M.remove(a,b)},_data:funct//...",
                "void 0===c?d&&\"get\"in d&&null!==(e=d.get(a,b))?e:(e=n.find.attr(a,b),null==//...",
                ";",
                "",
                "function sprintf() {",
                "/*"
            ],
            "filename": "jquery/jquery-2.1.4.min.js"
        },
        {
            "func": "$event.dispatch",
            "line": 374,
            "column": "2",
            "context": [
                "$event.dispatch = function( event ){",
                "\tif ( $.data( this, \"suppress.\"+ event.type ) - new Date().getTime() > 0 ){",
                "\t\t$.removeData( this, \"suppress.\"+ event.type );",
                "\t\treturn;",
                "\t}",
                "\treturn $dispatch.apply( this, arguments );",
                "};",
                "",
                "// event fix hooks for touch events...",
                "var touchHooks = ",
                "$event.fixHooks.touchstart = "
            ],
            "filename": "jquery/jquery.event.drag-2.2.js"
        },
        {
            "func": "r.handle",
            "line": 3,
            "column": "3176",
            "context": [
                "/*! jQuery v2.1.4 | (c) 2005, 2015 jQuery Foundation, Inc. | jquery.org/lic//...",
                "!function(a,b){\"object\"==typeof module&&\"object\"==typeof module.exports?mod//...",
                "return M.access(a,b,c)},removeData:function(a,b){M.remove(a,b)},_data:funct//...",
                "void 0===c?d&&\"get\"in d&&null!==(e=d.get(a,b))?e:(e=n.find.attr(a,b),null==//...",
                ";",
                "",
                "function sprintf() {",
                "/*"
            ],
            "filename": "jquery/jquery-2.1.4.min.js"
        }
    ],
    "useragent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.79 Safari/537.36 Edge/14.14393",
    "incomplete": "false",
    "partial": "true",
    "uri": "normalization.php?goto=tbl_structure.php&back=tbl_structure.php"
},
"script_name": "normalization.php",
"microhistory": null
```

**Comment by masoumii** [comment]
The bug still exists and we are on 4.8.5. now.... whats the deal with this?

**Comment by williamdes** [comment]
@masoumii Are you using a multi server setup ?

**Comment by masoumii** [comment]
> @masoumii Are you using a multi server setup ?

Yes, my hosting does indeed use directAdmin for it's server setup.
How can that cause this bug in PhpMyAdmin?

**Comment by williamdes** [comment]
@masoumii If you have multiple servers the server variable needs to be passed to the backend
I think I fixed this issue with 65deb85c34ca6b7366cf207e658c8163e7918439

Please download the latest version in development on our [website](https://www.phpmyadmin.net/downloads/) [(phpMyAdmin 4.8+snapshot)](https://www.phpmyadmin.net/downloads/) and comment if the latest version solved this issue :)
